src/main/kotlin/org/jetbrains/research/testspark/helpers/psiHelpers/JavaPsiHelper.kt

```diff
@@ -173,145 +136,313 @@ class JavaPsiHelper : PsiHelperInterface {
+                if (psiMethod.body == null) {
+                    return true
+                }
+            }
+
+            // check if a class is noted as abstract in the text
+            return psiClass.text.replace(" ", "")
+                .contains("abstractclass${psiClass.name}", ignoreCase = true)
+        }
+
     /**
      * Checks if the constraints on the selected class are satisfied, so that EvoSuite can generate tests for it.
      * Namely, it is not an enum and not an anonymous inner class.
      *
-     * @param psiClass the PSI class of interest
      * @return true if the constraints are satisfied, false otherwise
      */
-    private fun validateClass(psiClass: PsiClass): Boolean {
+    fun isValid(): Boolean {
         return !psiClass.isEnum && psiClass !is PsiAnonymousClass
     }
+}
 
-    /**
-     * Checks if the selected line contains (part of) a statement and is a valid line to be selected for EvoSuite.
-     * Namely, the line is within a method, it is not blank and contains (part of) a statement.
-     *
-     * @param selectedLine selected line number
-     * @param psiMethod surrounding PSI method
-     * @param psiFile containing PSI file
-     * @return true if the line is valid, false otherwise
-     */
-    private fun validateLine(
-        selectedLine: Int,
-        psiMethod: PsiMethod,
-        psiFile: PsiFile,
-    ): Boolean {
-        val doc: Document = PsiDocumentManager.getInstance(psiFile.project).getDocument(psiFile) ?: return false
+class JavaPsiHelper(private val psiFile: PsiFile) : PsiHelper {
 
-        val psiMethodBody: PsiCodeBlock = psiMethod.body ?: return false
-        if (psiMethodBody.statements.isEmpty()) return false
+    override val language: String
+        get() = "Java"
 
-        val firstStatement: PsiStatement = psiMethodBody.statements.first()
-        val lastStatement: PsiStatement = psiMethodBody.statements.last()
+    private val log = Logger.getInstance(this::class.java)
 
-        val firstStatementLine: Int = doc.getLineNumber(firstStatement.startOffset)
-        val lastStatementLine: Int = doc.getLineNumber(lastStatement.endOffset)
+    override fun generateMethodDescriptor(
+        psiMethod: PsiMethodWrapper,
+    ): String {
+        val methodDescriptor = psiMethod.methodDescriptor
+        log.info("Method description: $methodDescriptor")
+        return methodDescriptor
+    }
 
-        return (selectedLine in firstStatementLine..lastStatementLine)
+    override fun getSurroundingClass(
+        caretOffset: Int,
+    ): PsiClassWrapper? {
+        val classElements = PsiTreeUtil.findChildrenOfAnyType(psiFile, PsiClass::class.java)
+        for (cls in classElements) {
+            if (withinElement(cls, caretOffset)) {
+                val javaClassWrapper = JavaPsiClassWrapper(cls)
+                if (javaClassWrapper.isValid()) {
+                    log.info("Surrounding class for caret in $caretOffset is ${javaClassWrapper.qualifiedName}")
```

Comment:
Yeah, the name `isValid` for a method is not very good; I cannot understand what is checked without revising the implementation.